src/mono/mono/metadata/marshal-ilgen.c

```diff
@@ -29,18 +29,21 @@ static void emit_string_free_icall (MonoMethodBuilder *mb, MonoMarshalConv conv)
 // TODO: Does this need to loose the mono_ prefix?
 static void mono_marshal_ilgen_legacy_init (void);
 
-static gboolean ilgen_cb_inited = FALSE;
-static MonoMarshalIlgenCallbacks ilgen_marshal_cb;
+static MonoMarshalIlgenCallbacks* ilgen_marshal_cb = NULL;
 
 void
 mono_install_marshal_callbacks_ilgen (MonoMarshalIlgenCallbacks *cb)
 {
-	g_assert (!ilgen_cb_inited);
 	g_assert (cb->version == MONO_MARSHAL_CALLBACKS_VERSION);
-	memcpy (&ilgen_marshal_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
-	ilgen_cb_inited = TRUE;
-}
+	MonoMarshalIlgenCallbacks* local_cb = (MonoMarshalIlgenCallbacks*)malloc(sizeof(MonoMarshalIlgenCallbacks));
+	memcpy (local_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
```

Comment:
I think we need a memory barrier between the memcpy and the CAS

```suggestion
	memcpy (local_cb, cb, sizeof (MonoMarshalIlgenCallbacks));
	mono_memory_barrier ();
```